{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "# переконвертировать генбанк в gff\n",
    "\n",
    "from BCBio import GFF\n",
    "from Bio import SeqIO\n",
    "\n",
    "gbf_file = \"/mnt/projects/sochkalova/Psittacus/phyl/other_parrots.gbf\"\n",
    "gff_file = \"/mnt/projects/sochkalova/Psittacus/phyl/other_parrots.gff\"\n",
    "gbf_handle = open(gbf_file)\n",
    "gff_handle = open(gff_file, \"w\")\n",
    "\n",
    "GFF.write(SeqIO.parse(gbf_handle, \"genbank\"), gff_handle)\n",
    "\n",
    "gbf_handle.close()\n",
    "gff_handle.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      ">NC_045369.1 Agapornis lilianae\n",
      " 13 2\n",
      "\n",
      ">NC_011708.1 Agapornis roseicollis\n",
      " 13 2\n",
      "\n",
      ">NC_034679.1 Amazona ventralis\n",
      " 13 2\n",
      "\n",
      ">NC_029319.1 Ara ararauna\n",
      " 13 2\n",
      "\n",
      ">NC_045076.1 Ara macao\n",
      " 13 2\n",
      "\n",
      ">JX215256.1 Aratinga mitrata mitrata\n",
      " 13 2\n",
      "\n",
      ">NC_045371.1 Aratinga nenday\n",
      " 13 2\n",
      "\n",
      ">NC_026039.1 Aratinga solstitialis\n",
      " 13 2\n",
      "\n",
      ">NC_027841.1 Coracopsis vasa\n",
      " 13 2\n",
      "\n",
      ">NC_015197.1 Eupsittula pertinax chrysogenys\n",
      " 13 2\n",
      "\n",
      ">NC_026031.1 Guaruba guarouba\n",
      " 13 2\n",
      "\n",
      ">NC_040142.1 Cacatua pastinator\n",
      " 13 2\n",
      "\n",
      ">MF977813.1 Poicephalus gulielmi gulielmi\n",
      " 13 2\n",
      "\n",
      ">NC_044083.1 Poicephalus senegalus\n",
      " 13 2\n",
      "\n",
      ">NC_020325.1 Psittacara acuticaudatus acuticaudatus\n",
      " 13 2\n",
      "\n",
      ">NC_021764.1 Psittacara brevipes\n",
      " 13 2\n"
     ]
    }
   ],
   "source": [
    "# из gff непситтакусов сделать удобный для пользования файл с координатами генов\n",
    "\n",
    "ncbi_id = ''\n",
    "id_dict = {}\n",
    "counter = 0\n",
    "cds_amount = 0\n",
    "rna_amount = 0\n",
    "gff_file = \"/mnt/projects/sochkalova/Psittacus/phyl/other_parrots.gff\"\n",
    "out_file = \"/mnt/projects/sochkalova/Psittacus/phyl/cds_rna_other_parrots.tsv\"\n",
    "with open(gff_file) as infile:\n",
    "    with open(out_file, 'w') as outfile:\n",
    "        for line in infile:\n",
    "            line = line.strip().split('\\t')\n",
    "            if '##sequence-region' in line[0]:\n",
    "                if ncbi_id:\n",
    "                    header = '\\n>%s %s\\n' % (ncbi_id, species)\n",
    "                    print(header, end=' ')\n",
    "                    print(cds_amount, rna_amount)\n",
    "                    outfile.write(header)\n",
    "                    counter += 1\n",
    "                    ncbi_id = line[0].split()[1]\n",
    "                    cds_amount = 0\n",
    "                    rna_amount = 0\n",
    "                    for gene, coordinates in id_dict.items():\n",
    "                        s = '%s\\t%s\\t%s\\t%s\\n' % (coordinates.split()[0], coordinates.split()[1], gene.split()[0], gene.split()[1])\n",
    "                        outfile.write(s)\n",
    "                    id_dict = {}\n",
    "                    continue\n",
    "                    \n",
    "                if not ncbi_id:\n",
    "                    ncbi_id = line[0].split()[1]\n",
    "                    continue\n",
    "                    \n",
    "            if line[0] == ncbi_id:\n",
    "                if line[1] == 'annotation':\n",
    "                    if ncbi_id.startswith('NC') or ncbi_id.startswith('MG'):\n",
    "                        species = line[8].split(';')[6].replace('organism=','')\n",
    "                        continue\n",
    "                    else:\n",
    "                        species = line[8].split(';')[5].replace('organism=','')\n",
    "                        continue\n",
    "                    \n",
    "                if line[2] == 'rRNA':\n",
    "                    gene_name = 'rRNA {}'.format(line[8][-17:].replace(' ','_'))\n",
    "                    if gene_name not in id_dict.keys():\n",
    "                        id_dict[gene_name] = \"%s %s\" % (line[3], line[4])\n",
    "                        rna_amount += 1\n",
    "                        continue\n",
    "                        \n",
    "                if line[2] == 'CDS':\n",
    "                    if ncbi_id.startswith('NC'):\n",
    "                        gene_name = 'CDS {}'.format(line[8].split(';')[2].replace('gene=',''))\n",
    "                        if gene_name not in id_dict.keys():\n",
    "                            id_dict[gene_name] = \"%s %s\" % (line[3], line[4])\n",
    "                            cds_amount += 1\n",
    "                            continue\n",
    "                    else:\n",
    "                        gene_name = 'CDS {}'.format(line[8].split(';')[1].replace('gene=',''))\n",
    "                        if gene_name not in id_dict.keys():\n",
    "                            id_dict[gene_name] = \"%s %s\" % (line[3], line[4])\n",
    "                            cds_amount += 1\n",
    "                            continue\n",
    "            \n",
    "        if ncbi_id:\n",
    "            header = '\\n>%s %s\\n' % (ncbi_id, species)\n",
    "            print(header, end=' ')\n",
    "            print(cds_amount, rna_amount)\n",
    "            outfile.write(header)\n",
    "            counter += 1\n",
    "            for gene, coordinates in id_dict.items():\n",
    "                s = '%s\\t%s\\t%s\\t%s\\n' % (coordinates.split()[0], coordinates.split()[1], gene.split()[0], gene.split()[1])\n",
    "                outfile.write(s)\n",
    "                        \n",
    "#print(counter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">NC_045369.1 Agapornis lilianae\n",
      ">NC_011708.1 Agapornis roseicollis\n",
      ">NC_034679.1 Amazona ventralis\n",
      ">NC_029319.1 Ara ararauna\n",
      ">NC_045076.1 Ara macao\n",
      ">JX215256.1 Aratinga mitrata mitrata\n",
      ">NC_045371.1 Aratinga nenday\n",
      ">NC_026039.1 Aratinga solstitialis\n",
      ">NC_027841.1 Coracopsis vasa\n",
      ">NC_015197.1 Eupsittula pertinax chrysogenys\n",
      ">NC_026031.1 Guaruba guarouba\n",
      ">NC_040142.1 Cacatua pastinator\n",
      ">MF977813.1 Poicephalus gulielmi gulielmi\n",
      ">NC_044083.1 Poicephalus senegalus\n",
      ">NC_020325.1 Psittacara acuticaudatus acuticaudatus\n",
      ">NC_021764.1 Psittacara brevipes\n",
      ">Psittacus2\n",
      ">Psittacus1\n",
      ">Psittacus5\n",
      ">Psittacus10\n",
      ">Psittacus11_hyb\n",
      ">Psittacus9\n",
      ">Psittacus8\n",
      ">Psittacus12\n",
      ">Psittacus15\n"
     ]
    }
   ],
   "source": [
    "# распарсить мультифасты с непситтакусами и пситтакусами и сложить их в один словарь\n",
    "\n",
    "fasta = {} \n",
    "fasta_file = '/mnt/projects/sochkalova/Psittacus/phyl/other_parrots.mfa' \n",
    "header = None                             \n",
    "with open(fasta_file) as fh:              \n",
    "    for i, line in enumerate(fh):         \n",
    "        line = line.strip()               \n",
    "        if line.startswith(\">\"):          \n",
    "            if header:                    \n",
    "                fasta[header] = \"\".join(seq) \n",
    "            header = line             \n",
    "            seq = []                      \n",
    "        else:\n",
    "            seq.append(line)              \n",
    "    if header:                            \n",
    "        fasta[header] = \"\".join(seq)   \n",
    "\n",
    "\n",
    "fasta_file = '/mnt/projects/sochkalova/Psittacus/phyl/Psittacus_all.mfa'          \n",
    "header = None                             \n",
    "with open(fasta_file) as fh:              \n",
    "    for i, line in enumerate(fh):         \n",
    "        line = line.strip()               \n",
    "        if line.startswith(\">\"):          \n",
    "            if header:                    \n",
    "                fasta[header] = \"\".join(seq) \n",
    "            header = line             \n",
    "            seq = []                      \n",
    "        else:\n",
    "            seq.append(line)              \n",
    "    if header:                            \n",
    "        fasta[header] = \"\".join(seq)  \n",
    "        \n",
    "print(*fasta, sep='\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      ">Psittacus10\n",
      "\n",
      "13 2\n",
      "\n",
      ">Psittacus11_hyb\n",
      "\n",
      "13 2\n",
      "\n",
      ">Psittacus12\n",
      "\n",
      "13 2\n",
      "\n",
      ">Psittacus15\n",
      "\n",
      "13 2\n",
      "\n",
      ">Psittacus1\n",
      "\n",
      "13 2\n",
      "\n",
      ">Psittacus2\n",
      "\n",
      "13 2\n",
      "\n",
      ">Psittacus5\n",
      "\n",
      "13 2\n",
      "\n",
      ">Psittacus8\n",
      "\n",
      "13 2\n",
      "\n",
      ">Psittacus9\n",
      "\n",
      "13 2\n",
      "9\n"
     ]
    }
   ],
   "source": [
    "# дозаписать в файл с координатами генов координаты для пситтакусов\n",
    "ncbi_id = ''\n",
    "id_dict = {}\n",
    "counter = 0\n",
    "cds_amount = 0\n",
    "rna_amount = 0\n",
    "gff_file = \"/mnt/projects/sochkalova/Psittacus/phyl/Psittacus_all.gff\"\n",
    "out_file = \"/mnt/projects/sochkalova/Psittacus/phyl/cds_rna_other_parrots.tsv\"\n",
    "with open(gff_file) as infile:\n",
    "    with open(out_file, 'a') as outfile:\n",
    "        for line in infile:\n",
    "            line = line.strip().split('\\t')\n",
    "            \n",
    "            if line[0].startswith('>') and ncbi_id:\n",
    "                header = '\\n>%s\\n' % (ncbi_id)\n",
    "                print(header)\n",
    "                print(cds_amount, rna_amount)\n",
    "                outfile.write(header)\n",
    "                counter += 1\n",
    "                ncbi_id = line[0][1:]\n",
    "                cds_amount = 0\n",
    "                rna_amount = 0\n",
    "                for gene, coordinates in id_dict.items():\n",
    "                    s = '%s\\t%s\\t%s\\t%s\\n' % (coordinates.split()[0], coordinates.split()[1], gene.split()[0], gene.split()[1])\n",
    "                    outfile.write(s)\n",
    "                id_dict = {}\n",
    "                continue\n",
    "                    \n",
    "            if line[0].startswith('>') and not ncbi_id:\n",
    "                ncbi_id = line[0][1:]\n",
    "                continue\n",
    "                    \n",
    "            if line[0] == ncbi_id:                  \n",
    "                if line[2] == 'rRNA':\n",
    "                    gene_name = 'rRNA {}'.format(line[8].replace('Name=rrnS','12S_ribosomal_RNA').replace('Name=rrnL','16S_ribosomal_RNA'))\n",
    "                    if gene_name not in id_dict.keys():\n",
    "                        id_dict[gene_name] = \"%s %s\" % (line[3], line[4])\n",
    "                        rna_amount += 1\n",
    "                        continue\n",
    "                        \n",
    "                if line[2] == 'gene':\n",
    "                    gene_name = 'CDS {}'.format(line[8].replace('Name=',''))\n",
    "                    if gene_name not in id_dict.keys():\n",
    "                        id_dict[gene_name] = \"%s %s\" % (line[3], line[4])\n",
    "                        cds_amount += 1\n",
    "                        continue\n",
    "                        \n",
    "        if ncbi_id:\n",
    "            header = '\\n>%s\\n' % (ncbi_id)\n",
    "            print(header)\n",
    "            print(cds_amount, rna_amount)\n",
    "            outfile.write(header)\n",
    "            counter += 1\n",
    "            for gene, coordinates in id_dict.items():\n",
    "                s = '%s\\t%s\\t%s\\t%s\\n' % (coordinates.split()[0], coordinates.split()[1], gene.split()[0], gene.split()[1])\n",
    "                outfile.write(s)\n",
    "                        \n",
    "print(counter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/projects/sochkalova/Psittacus/phyl/genes/12S_ribosomal_RNA.mfa\n",
      "/mnt/projects/sochkalova/Psittacus/phyl/genes/16S_ribosomal_RNA.mfa\n",
      "/mnt/projects/sochkalova/Psittacus/phyl/genes/ATP6.mfa\n",
      "/mnt/projects/sochkalova/Psittacus/phyl/genes/ATP8.mfa\n",
      "/mnt/projects/sochkalova/Psittacus/phyl/genes/COX1.mfa\n",
      "/mnt/projects/sochkalova/Psittacus/phyl/genes/COX2.mfa\n",
      "/mnt/projects/sochkalova/Psittacus/phyl/genes/COX3.mfa\n",
      "/mnt/projects/sochkalova/Psittacus/phyl/genes/CYTB.mfa\n",
      "/mnt/projects/sochkalova/Psittacus/phyl/genes/ND1.mfa\n",
      "/mnt/projects/sochkalova/Psittacus/phyl/genes/ND2.mfa\n",
      "/mnt/projects/sochkalova/Psittacus/phyl/genes/ND3.mfa\n",
      "/mnt/projects/sochkalova/Psittacus/phyl/genes/ND4.mfa\n",
      "/mnt/projects/sochkalova/Psittacus/phyl/genes/ND4L.mfa\n",
      "/mnt/projects/sochkalova/Psittacus/phyl/genes/ND5.mfa\n",
      "/mnt/projects/sochkalova/Psittacus/phyl/genes/ND6.mfa\n"
     ]
    }
   ],
   "source": [
    "# для каждого гена в списке создать отдельный файл где будет последовательность нуклеотидов этого гена каждого попугая\n",
    "\n",
    "coordinates_file = \"/mnt/projects/sochkalova/Psittacus/phyl/cds_rna_other_parrots.tsv\"\n",
    "genes_list = ['12S_ribosomal_RNA', '16S_ribosomal_RNA', 'ATP6', 'ATP8', 'COX1', 'COX2', 'COX3', 'CYTB','ND1', 'ND2', 'ND3', 'ND4', 'ND4L', 'ND5', 'ND6']\n",
    "for gene in genes_list:\n",
    "    #print('\\n'+gene)\n",
    "    _id = ''\n",
    "    counter = 0\n",
    "    gene_output_file = \"/mnt/projects/sochkalova/Psittacus/phyl/genes/%s.mfa\" % gene\n",
    "    print(gene_output_file)\n",
    "    with open(gene_output_file, 'w') as output:\n",
    "         with open(coordinates_file) as fh:\n",
    "            for line in fh:\n",
    "                line = line.strip()\n",
    "                if line.startswith('>'):\n",
    "                    _id = line\n",
    "                    continue\n",
    "                if len(line.split('\\t')) > 3:\n",
    "                    if line.split('\\t')[3] == gene:\n",
    "                        start = int(line.split('\\t')[0])\n",
    "                        end = int(line.split('\\t')[1])\n",
    "                        seq_frag = fasta[_id][start:end+1]\n",
    "                        header = '\\n>%s\\n' % (_id[1:]) \n",
    "                        output.write(header)\n",
    "                        counter += 1\n",
    "                        print(counter, _id, sep=' ')\n",
    "                        output.write(seq_frag)\n",
    "                    else:\n",
    "                        continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#script to remove fucking spaces from gblocks output\n",
    "\n",
    "for gene in genes_list:\n",
    "    gblocks_output = \"/mnt/projects/sochkalova/Psittacus/phyl/genes/%s_aligned.mfa-gb\" % gene\n",
    "    no_gaps_file = \"/mnt/projects/sochkalova/Psittacus/phyl/genes/%s_aligned_conserv.mfa\" % gene\n",
    "    with open(gblocks_output) as ff:\n",
    "        with open(no_gaps_file, 'w') as fh:\n",
    "            for line in ff:\n",
    "                line = line.strip()\n",
    "                if line.startswith('>'):\n",
    "                    fh.write('\\n'+line+'\\n')\n",
    "                else:\n",
    "                    line = line.split()\n",
    "                    for block in line:\n",
    "                        fh.write(block)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      ">NC_045369.1 Agapornis lilianae\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 \n",
      ">NC_011708.1 Agapornis roseicollis\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 \n",
      ">NC_034679.1 Amazona ventralis\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 \n",
      ">NC_029319.1 Ara ararauna\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 \n",
      ">NC_045076.1 Ara macao\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 \n",
      ">JX215256.1 Aratinga mitrata mitrata\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 \n",
      ">NC_045371.1 Aratinga nenday\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 \n",
      ">NC_026039.1 Aratinga solstitialis\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 \n",
      ">NC_027841.1 Coracopsis vasa\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 \n",
      ">NC_015197.1 Eupsittula pertinax chrysogenys\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 \n",
      ">NC_026031.1 Guaruba guarouba\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 \n",
      ">NC_040142.1 Cacatua pastinator\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 \n",
      ">MF977813.1 Poicephalus gulielmi gulielmi\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 \n",
      ">NC_044083.1 Poicephalus senegalus\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 \n",
      ">NC_020325.1 Psittacara acuticaudatus acuticaudatus\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 \n",
      ">NC_021764.1 Psittacara brevipes\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 \n",
      ">Psittacus2\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 \n",
      ">Psittacus1\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 \n",
      ">Psittacus5\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 \n",
      ">Psittacus10\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 \n",
      ">Psittacus11_hyb\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 \n",
      ">Psittacus9\n",
      "12S_ribosomal_RNA 16S_ribosomal_RNA ATP6 ATP8 COX1 COX2 COX3 CYTB ND1 ND2 ND3 ND4 ND4L ND5 ND6 \n",
      ">Psittacus8\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 \n",
      ">Psittacus12\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 \n",
      ">Psittacus15\n",
      "12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 "
     ]
    }
   ],
   "source": [
    "# concatenation script\n",
    "\n",
    "final_file = '/mnt/projects/sochkalova/Psittacus/phyl/rrn_cds_aligned_concatenated.mfa' \n",
    "with open(final_file,  'w') as file:\n",
    "    for bird_id in fasta:\n",
    "        header = '\\n\\n%s\\n' % (bird_id.replace(' ', '_'))\n",
    "        file.write(header)\n",
    "        print()\n",
    "        print(bird_id)\n",
    "        gene_len = 0\n",
    "        for gene in genes_list:\n",
    "            header = None\n",
    "            gene_file = \"/mnt/projects/sochkalova/Psittacus/phyl/genes/%s_aligned_conserv.mfa\" % gene\n",
    "            with open(gene_file) as fh:\n",
    "                for line in fh:\n",
    "                    line = line.strip()\n",
    "                    if line == bird_id:\n",
    "                        header = True\n",
    "                        #print(line)\n",
    "                        continue\n",
    "                    if header:\n",
    "                        if line.startswith('>'):\n",
    "                            print(gene_len, end = ' ')\n",
    "                            break\n",
    "                        print(gene, end = ' ')\n",
    "                        file.write(line)\n",
    "                        gene_len += len(line)\n",
    "                        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12S_ribosomal_RNA = 1-954;\n",
      "16S_ribosomal_RNA = 955-2487;\n",
      "ATP6 = 2488-3168;\n",
      "ATP8 = 3169-3328;\n",
      "COX1 = 3329-4867;\n",
      "COX2 = 4868-5542;\n",
      "COX3 = 5543-6325;\n",
      "CYTB = 6326-7459;\n",
      "ND1 = 7460-8398;\n",
      "ND2 = 8399-9424;\n",
      "ND3 = 9425-9773;\n",
      "ND4 = 9774-11156;\n",
      "ND4L = 11157-11450;\n",
      "ND5 = 11451-13257;\n",
      "ND6 = 13258-13770;\n"
     ]
    }
   ],
   "source": [
    "string = '0 12S_ribosomal_RNA 954 16S_ribosomal_RNA 2487 ATP6 3168 ATP8 3328 COX1 4867 COX2 5542 COX3 6325 CYTB 7459 ND1 8398 ND2 9424 ND3 9773 ND4 11156 ND4L 11450 ND5 13257 ND6 13770 '\n",
    "string = string.split()\n",
    "for i in range(len(string)):\n",
    "    if i != 0 and string[i].isdigit():\n",
    "        start = int(string[i-2])+1\n",
    "        end = int(string[i])\n",
    "        line = '%s = %d-%d;' % (string[i-1], start, end)\n",
    "        print(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
